## Data Wrangling with Python Project_DE_Nelson_Otuma_Ongaya_2021
Project Deliverable
- Your deliverable will be a python notebook that will contain your solution.
- You will need to submit the shareable link to your notebook

## Problem Statement

As a Data Science Consultant, you need to provide recommendations on reducing the cases of
breakdown of buses in the city of New York.
The dataset you have provided comes from the Bus Breakdown and Delay system, which
collects information from school bus vendors operating in the field in real-time. Bus staff that
encounter delays during the route are instructed to radio the dispatcher at the bus vendor’s
central office. The bus vendor staff log into the Bus Breakdown and Delay system to record the
event and notify OPT. OPT customer service agents use this system to inform parents who call
with questions regarding bus service. The Bus Breakdown and Delay system is publicly
accessible and contains real-time updates. All information in the system is by school bus vendor
staff.
You start this project by reading, exploring, cleaning, and later, performing an analysis of your
dataset.

#### Below are the Research Quetions that we need to answer as part of our analysis for this project
1. Which bus companies had the highest breakdowns?
2. What were the top 3 reasons for bus delays?
3. How many students were on the buses when they broke down?
4. Which were the most frequent reasons for bus breakdowns?
5. What were the most frequent reasons for the bus running late?
6. What was the average delay time of each reason type?

## 1. Data Understanding

In [1]:
# we import the libraries for reading our dataset and analysis
import pandas as pd
import numpy as np

In [2]:
df=pd.read_csv('bus-breakdown-and-delays.csv')#read the dataset

In [3]:
df.head()#preview the first five records in our dataset.

School_Year  Busbreakdown_ID           Run_Type Bus_No Route_Number  \
0   2015-2016          1227538  Special Ed AM Run   2621         J711   
1   2015-2016          1227539  Special Ed AM Run   1260         M351   
2   2015-2016          1227540           Pre-K/EI    418            3   
3   2015-2016          1227541  Special Ed AM Run   4522         M271   
4   2015-2016          1227542  Special Ed AM Run   3124         M373   

          Reason Schools_Serviced     Occurred_On      Created_On        Boro  \
0  Heavy Traffic            75003  11/5/2015 8:10  11/5/2015 8:12  New Jersey   
1  Heavy Traffic             6716  11/5/2015 8:10  11/5/2015 8:12   Manhattan   
2  Heavy Traffic             C445  11/5/2015 8:09  11/5/2015 8:13       Bronx   
3  Heavy Traffic             2699  11/5/2015 8:12  11/5/2015 8:14   Manhattan   
4  Heavy Traffic             2116  11/5/2015 8:13  11/5/2015 8:14   Manhattan   

   ... How_Long_Delayed Number_Of_Students_On_The_Bus  \
0  ...              NaN                            11   
1  ...            20MNS                             2   
2  ...            15MIN                             8   
3  ...           15 MIN                             6   
4  ...              NaN                             6   

   Has_Contractor_Notified_Schools Has_Contractor_Notified_Parents  \
0                              Yes                              No   
1                              Yes                             Yes   
2                              Yes                             Yes   
3                               No                              No   
4                               No                              No   

  Have_You_Alerted_OPT     Informed_On Incident_Number Last_Updated_On  \
0                  Yes  11/5/2015 8:12             NaN  11/5/2015 8:12   
1                   No  11/5/2015 8:12             NaN  11/5/2015 8:13   
2                  Yes  11/5/2015 8:13             NaN  11/5/2015 8:13   
3                   No  11/5/2015 8:14             NaN  11/5/2015 8:14   
4                   No  11/5/2015 8:14             NaN  11/5/2015 8:14   

  Breakdown_or_Running_Late School_Age_or_PreK  
0              Running Late         School-Age  
1              Running Late         School-Age  
2              Running Late              Pre-K  
3              Running Late         School-Age  
4              Running Late         School-Age  

[5 rows x 21 columns]

In [4]:
df.shape#preview the no. of records

(281110, 21)

--The bus -breakdown-and-delay dataset has total 281110 records with 21 columns

In [5]:
#Previewing the last few records
df.tail()

School_Year  Busbreakdown_ID             Run_Type  Bus_No Route_Number  \
281105   2016-2017          1338452             Pre-K/EI    9345            2   
281106   2016-2017          1341521             Pre-K/EI       1            5   
281107   2016-2017          1353044    Special Ed PM Run  GC0112         X928   
281108   2016-2017          1353045    Special Ed PM Run   5525D         Q920   
281109   2016-2017          1353046  Project Read PM Run    2530         K617   

               Reason Schools_Serviced      Occurred_On       Created_On  \
281105  Heavy Traffic             C530    4/5/2017 8:00    4/5/2017 8:10   
281106  Heavy Traffic             C579   4/24/2017 7:42   4/24/2017 7:44   
281107  Heavy Traffic             9003  5/25/2017 16:22  5/25/2017 16:28   
281108    Won`t Start            24457  5/25/2017 16:27  5/25/2017 16:30   
281109          Other            21436  5/25/2017 16:36  5/25/2017 16:37   

            Boro  ... How_Long_Delayed Number_Of_Students_On_The_Bus  \
281105     Bronx  ...            15-20                             7   
281106     Bronx  ...          20 MINS                             0   
281107     Bronx  ...        20-25MINS                             0   
281108    Queens  ...              NaN                             0   
281109  Brooklyn  ...            45min                             7   

        Has_Contractor_Notified_Schools Has_Contractor_Notified_Parents  \
281105                              Yes                             Yes   
281106                              Yes                             Yes   
281107                              Yes                             Yes   
281108                              Yes                             Yes   
281109                              Yes                             Yes   

       Have_You_Alerted_OPT      Informed_On Incident_Number  Last_Updated_On  \
281105                   No    4/5/2017 8:10             NaN    4/5/2017 8:10   
281106                   No   4/24/2017 7:44             NaN   4/24/2017 7:44   
281107                  Yes  5/25/2017 16:28        90323827  5/25/2017 16:34   
281108                   No  5/25/2017 16:30             NaN  5/25/2017 16:30   
281109                  Yes  5/25/2017 16:37             NaN  5/25/2017 16:37   

       Breakdown_or_Running_Late School_Age_or_PreK  
281105              Running Late              Pre-K  
281106              Running Late              Pre-K  
281107              Running Late         School-Age  
281108                 Breakdown         School-Age  
281109              Running Late         School-Age  

[5 rows x 21 columns]

In [6]:
#Preview a sample of 10 records from the  dataset
df.sample(10)

School_Year  Busbreakdown_ID           Run_Type   Bus_No Route_Number  \
121503   2018-2019          1482388  Special Ed AM Run    10483         X361   
82439    2016-2017          1331495  General Ed AM Run    26627        K2150   
30152    2015-2016          1268937  Special Ed PM Run     1495         M293   
92208    2016-2017          1342785  Special Ed AM Run   TN0344         Q619   
95955    2016-2017          1347155  Special Ed AM Run     1338         M650   
21522    2015-2016          1259345  Special Ed AM Run  VN24641         X185   
266006   2018-2019          1476993  Special Ed AM Run     1644         M992   
149268   2017-2018          1376999  General Ed AM Run     1661        Q2867   
156354   2017-2018          1370170  Special Ed AM Run     1239         M558   
58876    2016-2017          1303958  Special Ed AM Run     599D         Q553   

                    Reason    Schools_Serviced      Occurred_On  \
121503       Heavy Traffic  80,620,813,008,333   11/8/2018 6:50   
82439        Heavy Traffic               21730   3/17/2017 7:20   
30152        Heavy Traffic                3448  5/17/2016 13:54   
92208        Heavy Traffic               30008   4/27/2017 6:05   
95955        Heavy Traffic                2605   5/10/2017 7:45   
21522        Heavy Traffic                8375    4/1/2016 6:19   
266006       Heavy Traffic  31,990,345,203,811  10/26/2018 7:10   
149268  Mechanical Problem       2,834,228,773  10/16/2017 6:30   
156354       Heavy Traffic         341,703,802   9/26/2017 7:10   
58876        Heavy Traffic               30008   12/1/2016 7:04   

             Created_On       Boro  ... How_Long_Delayed  \
121503   11/8/2018 6:59      Bronx  ...        31-45 Min   
82439    3/17/2017 7:34   Brooklyn  ...          10 mins   
30152   5/17/2016 13:54  Manhattan  ...          20 MINS   
92208    4/27/2017 6:17     Queens  ...          30mins.   
95955    5/10/2017 7:45  Manhattan  ...          20 MINS   
21522     4/1/2016 6:21      Bronx  ...           25mins   
266006  10/26/2018 7:12  Manhattan  ...        31-45 Min   
149268  10/16/2017 8:01     Queens  ...              NaN   
156354   9/26/2017 7:11  Manhattan  ...        46-60 Min   
58876    12/1/2016 7:05     Queens  ...           30MINS   

       Number_Of_Students_On_The_Bus  Has_Contractor_Notified_Schools  \
121503                             0                              Yes   
82439                              0                              Yes   
30152                              0                              Yes   
92208                              0                              Yes   
95955                              6                              Yes   
21522                              0                              Yes   
266006                             0                              Yes   
149268                             0                              Yes   
156354                             1                              Yes   
58876                              0                              Yes   

       Has_Contractor_Notified_Parents Have_You_Alerted_OPT      Informed_On  \
121503                             Yes                   No   11/8/2018 6:59   
82439                               No                  Yes   3/17/2017 7:34   
30152                              Yes                   No  5/17/2016 13:54   
92208                              Yes                  Yes   4/27/2017 6:17   
95955                              Yes                   No   5/10/2017 7:45   
21522                              Yes                  Yes    4/1/2016 6:21   
266006                             Yes                   No  10/26/2018 7:12   
149268                              No                   No  10/16/2017 8:01   
156354                             Yes                   No   9/26/2017 7:11   
58876                              Yes                   No   12/1/2016 7:05   

       Incident_Number  Last_Updated_On Br

In [7]:
# Checking the datatypes of df variables(columns)
df.dtypes

School_Year                        object
Busbreakdown_ID                     int64
Run_Type                           object
Bus_No                             object
Route_Number                       object
Reason                             object
Schools_Serviced                   object
Occurred_On                        object
Created_On                         object
Boro                               object
Bus_Company_Name                   object
How_Long_Delayed                   object
Number_Of_Students_On_The_Bus       int64
Has_Contractor_Notified_Schools    object
Has_Contractor_Notified_Parents    object
Have_You_Alerted_OPT               object
Informed_On                        object
Incident_Number                    object
Last_Updated_On                    object
Breakdown_or_Running_Late          object
School_Age_or_PreK                 object
dtype: object

In [8]:
#check for Index range,data columns and entries in our dataset
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 281110 entries, 0 to 281109
Data columns (total 21 columns):
 #   Column                           Non-Null Count   Dtype 
---  ------                           --------------   ----- 
 0   School_Year                      281110 non-null  object
 1   Busbreakdown_ID                  281110 non-null  int64 
 2   Run_Type                         281107 non-null  object
 3   Bus_No                           281101 non-null  object
 4   Route_Number                     281103 non-null  object
 5   Reason                           281108 non-null  object
 6   Schools_Serviced                 281103 non-null  object
 7   Occurred_On                      281110 non-null  object
 8   Created_On                       281110 non-null  object
 9   Boro                             267649 non-null  object
 10  Bus_Company_Name                 281110 non-null  object
 11  How_Long_Delayed                 245502 non-null  object
 12  Number_Of_Studen

- The dataset has 281110 entries, 21 column variables, dtype, int64 and object(19)

## 2. Data Cleaning

- We clean the data set by checking for any iconsistent column names, missing data in the dataset and duplicate values

In [9]:
# we check for the missing values and in our data set
df.isnull().sum()

School_Year                             0
Busbreakdown_ID                         0
Run_Type                                3
Bus_No                                  9
Route_Number                            7
Reason                                  2
Schools_Serviced                        7
Occurred_On                             0
Created_On                              0
Boro                                13461
Bus_Company_Name                        0
How_Long_Delayed                    35608
Number_Of_Students_On_The_Bus           0
Has_Contractor_Notified_Schools         0
Has_Contractor_Notified_Parents         0
Have_You_Alerted_OPT                    0
Informed_On                             0
Incident_Number                    271627
Last_Updated_On                         0
Breakdown_or_Running_Late               0
School_Age_or_PreK                      0
dtype: int64

We note that the variables have missing data entries.
- Incident_Number  271627
- How_Long_Delayed 35608
- Boro             13461
- Schools_Service      7
- Reason               2
- Route_Number         7
- Run_Type             3
- Bus_No               9
The incident_number should be recorded on the data set as a trucking number from the OPT Customer Service
also we note that the OPT Customer Service has not taken into consideration the variable How_Long_Delayed. Boro as a variable refers to the state that holds the subtowns.

In [10]:
#Checking for duplicates in our dataset
sum(df.duplicated())

0

+ we do not have duplicated values in our dataset.

In [11]:
# we check if any of the columns are all null
df.isnull().all()

School_Year                        False
Busbreakdown_ID                    False
Run_Type                           False
Bus_No                             False
Route_Number                       False
Reason                             False
Schools_Serviced                   False
Occurred_On                        False
Created_On                         False
Boro                               False
Bus_Company_Name                   False
How_Long_Delayed                   False
Number_Of_Students_On_The_Bus      False
Has_Contractor_Notified_Schools    False
Has_Contractor_Notified_Parents    False
Have_You_Alerted_OPT               False
Informed_On                        False
Incident_Number                    False
Last_Updated_On                    False
Breakdown_or_Running_Late          False
School_Age_or_PreK                 False
dtype: bool

++ We note that the records have observations recorded in the the columns.

In [12]:
# We check if any of the rows in the dataset are all null
df.isnull().all(1).any()

False

+ none of the rows in the dataset are null

## Data Standardisation

***we standardize our dataset by fixing inconsistent column names. we can perform data type conversion to fix inaccurate data.

In [13]:
list(df.columns)

['School_Year',
 'Busbreakdown_ID',
 'Run_Type',
 'Bus_No',
 'Route_Number',
 'Reason',
 'Schools_Serviced',
 'Occurred_On',
 'Created_On',
 'Boro',
 'Bus_Company_Name',
 'How_Long_Delayed',
 'Number_Of_Students_On_The_Bus',
 'Has_Contractor_Notified_Schools',
 'Has_Contractor_Notified_Parents',
 'Have_You_Alerted_OPT',
 'Informed_On',
 'Incident_Number',
 'Last_Updated_On',
 'Breakdown_or_Running_Late',
 'School_Age_or_PreK']

All the variables are uniformly named, have followed standard capitalization 99 percent of the format. 

In [14]:
 ### We check  if the "Yes/No" fields contain only these 2 values
 ### for have_you_alerted_opt variable
# Hint: Use unique() function
data  = df['Have_You_Alerted_OPT'].unique()
list(data)

['Yes', 'No']

- The field contains Yes or No 

In [15]:
#Does the 'Yes'or 'No' field contain 2 values?
# for has_contractor_notified_parents variable
data1 = df['Has_Contractor_Notified_Parents'].unique()
list(data1)

['No', 'Yes']

+ We note that our Has_Contractor_Notified_Parents variable contains only two values.

In [16]:
# we check for unique values in break_down_or_running_late variable to ensure there is no duplication

data2 = df['Breakdown_or_Running_Late'].unique()
list(data2)

['Running Late', 'Breakdown']

+ We note that our 'Breakdown_or_Running_Late' variable contains only two values.'Running Late', 'Breakdown'

In [17]:
# check for unique values in school_age_or_prek variable
data3 = df['School_Age_or_PreK'].unique()
list(data3)

['School-Age', 'Pre-K']

+ Yes it has two unique values ['School-Age', 'Pre-K']

In [18]:
#Perform a unique check values in school_year variable
data4=df['School_Year'].unique()

In [19]:
list(data4)

['2015-2016', '2017-2018', '2018-2019', '2016-2017', '2019-2020']

In [20]:
# Check for unique values in reason variable
reason_df = df['Reason'].unique()
list(reason_df)

['Heavy Traffic',
 'Flat Tire',
 'Other',
 'Won`t Start',
 'Mechanical Problem',
 'Problem Run',
 'Accident',
 'Late return from Field Trip',
 'Delayed by School',
 'Weather Conditions',
 nan]

In [21]:
# Check unique values in run_type variable
runtype_df = df['Run_Type'].unique()
list(runtype_df)

['Special Ed AM Run',
 'Pre-K/EI',
 'General Ed AM Run',
 'General Ed Field Trip',
 'Special Ed PM Run',
 'General Ed PM Run',
 'Project Read PM Run',
 'Special Ed Field Trip',
 'Project Read AM Run',
 'Project Read Field Trip',
 nan]

In [22]:
#Check for unique values in boro variable
boro_df = df['Boro'].unique()
list(boro_df)

['New Jersey',
 'Manhattan',
 'Bronx',
 'Westchester',
 'Brooklyn',
 'Rockland County',
 'Nassau County',
 nan,
 'Queens',
 'Staten Island',
 'Connecticut',
 'All Boroughs']

## Overall Data Cleaning Observations

#### Missing Values

+ "How_Long_Delayed" and "Incident_number"  has the largest number of missing values that would help us in drawing corelations in our dataset.

####  Error in values

we need to convert "MINS" in the How_Long_Delayed" into integer

#### Error in Datatypes

+ we need to "How_Long_Delayed convert into integer

#### Error in field names

-We need to rename "Boro" to "Borough"

## 3. Handling Errors,Missing Values,Datatypes and Error in Fields.

In [23]:
#we create a new dataset from our original dataset.
df_clean = df.copy()
df_clean.sample()

School_Year  Busbreakdown_ID           Run_Type Bus_No Route_Number  \
27983   2015-2016          1266417  Special Ed AM Run   2671         M206   

              Reason Schools_Serviced    Occurred_On     Created_On  \
27983  Heavy Traffic             1134  5/6/2016 7:56  5/6/2016 7:59   

            Boro  ... How_Long_Delayed Number_Of_Students_On_The_Bus  \
27983  Manhattan  ...              NaN                             0   

       Has_Contractor_Notified_Schools Has_Contractor_Notified_Parents  \
27983                              Yes                             Yes   

      Have_You_Alerted_OPT    Informed_On Incident_Number Last_Updated_On  \
27983                  Yes  5/6/2016 7:59             NaN   5/6/2016 7:59   

      Breakdown_or_Running_Late School_Age_or_PreK  
27983              Running Late         School-Age  

[1 rows x 21 columns]

In [24]:
# Extract the lowest delay time in the column how_long_delayed from the string
# 
df_clean['How_Long_Delayed'] = df_clean['How_Long_Delayed'].str.extract('(\d+)')
df_clean['How_Long_Delayed'].head()

0    NaN
1     20
2     15
3     15
4    NaN
Name: How_Long_Delayed, dtype: object

In [25]:
# Covert How_Long_Delayed  into float.
df_clean['How_Long_Delayed'] = df_clean['How_Long_Delayed'].apply(np.float)
df_clean['How_Long_Delayed'].dtypes

<ipython-input-25-3278af16383f>:2: DeprecationWarning: `np.float` is a deprecated alias for the builtin `float`. To silence this warning, use `float` by itself. Doing this will not modify any behavior and is safe. If you specifically wanted the numpy scalar type, use `np.float64` here.
Deprecated in NumPy 1.20; for more details and guidance: https://numpy.org/devdocs/release/1.20.0-notes.html#deprecations
  df_clean['How_Long_Delayed'] = df_clean['How_Long_Delayed'].apply(np.float)


dtype('float64')

In [26]:
#we can use the mean to impute the missing values
missing_col = ['How_Long_Delayed']
#Technique 1: Using mean to impute the missing values
for i in missing_col:
 df_clean.loc[df_clean.loc[:,i].isnull(),i]=df_clean.loc[:,i].mean()
df_clean['How_Long_Delayed'].head()

0    28.187284
1    20.000000
2    15.000000
3    15.000000
4    28.187284
Name: How_Long_Delayed, dtype: float64

In [27]:
df_clean['How_Long_Delayed'].mean()

28.187284185448267

In [28]:
#convert back our how_long_delayed column to integer datatype
df_clean['How_Long_Delayed'] = df_clean['How_Long_Delayed'].apply(np.int)
df_clean['How_Long_Delayed'].dtypes

<ipython-input-28-a110602c8e62>:2: DeprecationWarning: `np.int` is a deprecated alias for the builtin `int`. To silence this warning, use `int` by itself. Doing this will not modify any behavior and is safe. When replacing `np.int`, you may wish to use e.g. `np.int64` or `np.int32` to specify the precision. If you wish to review your current use, check the release note link for additional information.
Deprecated in NumPy 1.20; for more details and guidance: https://numpy.org/devdocs/release/1.20.0-notes.html#deprecations
  df_clean['How_Long_Delayed'] = df_clean['How_Long_Delayed'].apply(np.int)


dtype('int64')

In [29]:
# Checking for nulls in the column
df_clean['How_Long_Delayed'].isnull().sum()

0

In [30]:
# we need to rename Boro column to Borough
df_clean.rename(columns={'Boro':'Borough'},inplace=True)
df_clean.columns

Index(['School_Year', 'Busbreakdown_ID', 'Run_Type', 'Bus_No', 'Route_Number',
       'Reason', 'Schools_Serviced', 'Occurred_On', 'Created_On', 'Borough',
       'Bus_Company_Name', 'How_Long_Delayed', 'Number_Of_Students_On_The_Bus',
       'Has_Contractor_Notified_Schools', 'Has_Contractor_Notified_Parents',
       'Have_You_Alerted_OPT', 'Informed_On', 'Incident_Number',
       'Last_Updated_On', 'Breakdown_or_Running_Late', 'School_Age_or_PreK'],
      dtype='object')

#### Converting all columns varaibles to lower case

In [31]:
df_clean.columns = df_clean.columns.str.lower()
df_clean.columns

Index(['school_year', 'busbreakdown_id', 'run_type', 'bus_no', 'route_number',
       'reason', 'schools_serviced', 'occurred_on', 'created_on', 'borough',
       'bus_company_name', 'how_long_delayed', 'number_of_students_on_the_bus',
       'has_contractor_notified_schools', 'has_contractor_notified_parents',
       'have_you_alerted_opt', 'informed_on', 'incident_number',
       'last_updated_on', 'breakdown_or_running_late', 'school_age_or_prek'],
      dtype='object')

In [32]:
#converting all columns into standard for easy reading.
df_clean = df_clean.applymap(lambda s: s.lower() if type(s) == str else s)
df_clean

school_year  busbreakdown_id             run_type  bus_no route_number  \
0        2015-2016          1227538    special ed am run    2621         j711   
1        2015-2016          1227539    special ed am run    1260         m351   
2        2015-2016          1227540             pre-k/ei     418            3   
3        2015-2016          1227541    special ed am run    4522         m271   
4        2015-2016          1227542    special ed am run    3124         m373   
...            ...              ...                  ...     ...          ...   
281105   2016-2017          1338452             pre-k/ei    9345            2   
281106   2016-2017          1341521             pre-k/ei       1            5   
281107   2016-2017          1353044    special ed pm run  gc0112         x928   
281108   2016-2017          1353045    special ed pm run   5525d         q920   
281109   2016-2017          1353046  project read pm run    2530         k617   

               reason schools_serviced      occurred_on       created_on  \
0       heavy traffic            75003   11/5/2015 8:10   11/5/2015 8:12   
1       heavy traffic             6716   11/5/2015 8:10   11/5/2015 8:12   
2       heavy traffic             c445   11/5/2015 8:09   11/5/2015 8:13   
3       heavy traffic             2699   11/5/2015 8:12   11/5/2015 8:14   
4       heavy traffic             2116   11/5/2015 8:13   11/5/2015 8:14   
...               ...              ...              ...              ...   
281105  heavy traffic             c530    4/5/2017 8:00    4/5/2017 8:10   
281106  heavy traffic             c579   4/24/2017 7:42   4/24/2017 7:44   
281107  heavy traffic             9003  5/25/2017 16:22  5/25/2017 16:28   
281108    won`t start            24457  5/25/2017 16:27  5/25/2017 16:30   
281109          other            21436  5/25/2017 16:36  5/25/2017 16:37   

           borough  ... how_long_delayed  number_of_students_on_the_bus  \
0       new jersey  ...               28                             11   
1        manhattan  ...               20                              2   
2            bronx  ...               15                              8   
3        manhattan  ...               15                              6   
4        manhattan  ...               28                              6   
...            ...  ...              ...                            ...   
281105       bronx  ...               15                              7   
281106       bronx  ...               20                              0   
281107       bronx  ...               20                              0   
281108      queens  ...               28                              0   
281109    brooklyn  ...               45                              7   

        has_contractor_notified_schools has_contractor_notified_parents  \
0                                   yes                              no   
1                                   yes                             yes   
2                                   yes                             yes   
3                                    no                              no   
4                                    no                              no   
...                                 ...                             ...   
281105                              yes                             yes   
281106                              yes                             yes   
281107                              yes                             yes   
281108                              yes                             yes   
281109                              yes                             yes   

       have_you_alerted_opt      informed_on incident_number  last_updated_on  \
0                       yes   11/5/2015 8:12             NaN   11/5/2015 8:12   
1                        no   11/5/2015 8:12             NaN   11/5/2015 8:13   
2                       yes   11/5/2015 8:13             NaN   11/5/2015 8:1

In [33]:
#Lets check our cleaned dataset.
df_clean.head()

school_year  busbreakdown_id           run_type bus_no route_number  \
0   2015-2016          1227538  special ed am run   2621         j711   
1   2015-2016          1227539  special ed am run   1260         m351   
2   2015-2016          1227540           pre-k/ei    418            3   
3   2015-2016          1227541  special ed am run   4522         m271   
4   2015-2016          1227542  special ed am run   3124         m373   

          reason schools_serviced     occurred_on      created_on     borough  \
0  heavy traffic            75003  11/5/2015 8:10  11/5/2015 8:12  new jersey   
1  heavy traffic             6716  11/5/2015 8:10  11/5/2015 8:12   manhattan   
2  heavy traffic             c445  11/5/2015 8:09  11/5/2015 8:13       bronx   
3  heavy traffic             2699  11/5/2015 8:12  11/5/2015 8:14   manhattan   
4  heavy traffic             2116  11/5/2015 8:13  11/5/2015 8:14   manhattan   

   ... how_long_delayed  number_of_students_on_the_bus  \
0  ...               28                             11   
1  ...               20                              2   
2  ...               15                              8   
3  ...               15                              6   
4  ...               28                              6   

   has_contractor_notified_schools has_contractor_notified_parents  \
0                              yes                              no   
1                              yes                             yes   
2                              yes                             yes   
3                               no                              no   
4                               no                              no   

  have_you_alerted_opt     informed_on incident_number last_updated_on  \
0                  yes  11/5/2015 8:12             NaN  11/5/2015 8:12   
1                   no  11/5/2015 8:12             NaN  11/5/2015 8:13   
2                  yes  11/5/2015 8:13             NaN  11/5/2015 8:13   
3                   no  11/5/2015 8:14             NaN  11/5/2015 8:14   
4                   no  11/5/2015 8:14             NaN  11/5/2015 8:14   

  breakdown_or_running_late school_age_or_prek  
0              running late         school-age  
1              running late         school-age  
2              running late              pre-k  
3              running late         school-age  
4              running late         school-age  

[5 rows x 21 columns]

## 4. Research Questions.

+ We now address our research questions since we already have cleaned dataset.

In [34]:
# (i). Which bus companies that had the highest breakdowns?
highest_breakdowns = pd.DataFrame(df_clean.groupby(['bus_company_name'])['busbreakdown_id'].count())
highest_breakdowns.sort_values(by='busbreakdown_id',ascending=0)

busbreakdown_id
bus_company_name                                   
g.v.c., ltd.                                  19394
leesel transportation corp (b2192)            17200
reliant trans, inc. (b232                     13741
pioneer transportation co                     12017
boro transit, inc.                            11953
...                                             ...
l&m bus corp.                                     1
fortuna bus company                               1
phillip bus service                               1
r & c transit, inc. (b2321)                       1
1967                                              1

[116 rows x 1 columns]

+ We note that G.V.C,LTD bus_company had the highest breakdown  of 19394

In [35]:
#(ii). What were the top 3 reasons for bus delays?
# We sort by frequency to get top 3 raesons for delays under reason
delay_df = pd.DataFrame(df_clean.groupby(['reason'])['busbreakdown_id'].count())
delay_df.sort_values(by = 'busbreakdown_id', ascending = 0).head(3)

busbreakdown_id
reason                             
heavy traffic                173221
other                         37579
mechanical problem            28162

+ Heavy Traffic was the Most common reason for bus delays followed by other then Mechanical Problem.

In [36]:
#(iii). How many students were in the buses when they broke down? 
#We apply the grouby concept.
#df_clean.groupby(['number_of_students_on_the_bus']).count()[['busbreakdown_id']].reset_index()
df_clean.groupby(['busbreakdown_id','breakdown_or_running_late']).agg({'number_of_students_on_the_bus': ['sum']})

number_of_students_on_the_bus
                                                                    sum
busbreakdown_id breakdown_or_running_late                              
1212681         running late                                          0
1212682         running late                                          0
1212683         running late                                          1
1212684         running late                                          0
1212685         running late                                          0
...                                                                 ...
1501608         running late                                          0
1501609         running late                                          0
1501610         running late                                          0
1501611         running late                                          1
1501612         running late                                          1

[281110 rows x 1 columns]

In [37]:
#(iv). Which were most frequent reasons for bus breakdowns?
breakdown_reasons = df_clean[df_clean.breakdown_or_running_late == 'breakdown'].groupby(['reason'])['busbreakdown_id'].count()
breakdown_reasons.sort_values(ascending=False)

reason
mechanical problem             14985
won`t start                     7731
flat tire                       4038
other                           3530
heavy traffic                    419
accident                         202
weather conditions                82
late return from field trip       35
problem run                       28
delayed by school                  7
Name: busbreakdown_id, dtype: int64

In [38]:
#++ We can also tell the bus company and reason for breakdown.
breakdown_reason = df_clean[df_clean.breakdown_or_running_late == 'breakdown'].groupby(['bus_company_name','reason'])['busbreakdown_id'].count()
breakdown_reason.sort_values(ascending=False)

bus_company_name           reason            
little richie bus service  won`t start           2045
                           mechanical problem    1763
logan bus company inc.     won`t start           1073
reliant trans, inc. (b232  mechanical problem    1043
little richie bus service  flat tire              961
                                                 ... 
selby transportation       flat tire                1
selby trans corp. (b2192)  other                    1
little richie bus service  problem run              1
logan bus company inc.     delayed by school        1
1992                       mechanical problem       1
Name: busbreakdown_id, Length: 498, dtype: int64

In [39]:
 #(v). What were the most frequent reasons for the bus running late?
# we need to get the records with running late reasons and sort to get most frequent reasons
reasons = df_clean.groupby(['reason','how_long_delayed']).count()
reasons.sort_values(by = 'how_long_delayed', ascending=0)


school_year  busbreakdown_id  \
reason                      how_long_delayed                                 
late return from field trip 202340                      1                1   
heavy traffic               56108                       1                1   
                            27832                       1                1   
                            26630                       1                1   
mechanical problem          25550                       1                1   
...                                                   ...              ...   
problem run                 0                          66               66   
flat tire                   0                         113              113   
weather conditions          0                         299              299   
delayed by school           0                         269              269   
accident                    0                          32               32   

                                              run_type  bus_no  route_number  \
reason                      how_long_delayed                                   
late return from field trip 202340                   1       1             1   
heavy traffic               56108                    1       1             1   
                            27832                    1       1             1   
                            26630                    1       1             1   
mechanical problem          25550                    1       1             1   
...                                                ...     ...           ...   
problem run                 0                       66      66            66   
flat tire                   0                      113     113           113   
weather conditions          0                      299     299           299   
delayed by school           0                      269     268           267   
accident                    0                       32      32            32   

                                              schools_serviced  occurred_on  \
reason                      how_long_delayed                                  
late return from field trip 202340                           1            1   
heavy traffic               56108                            1            1   
                            27832                            1            1   
                            26630                            1            1   
mechanical problem          25550                            1            1   
...                                                        ...          ...   
problem run                 0                               66           66   
flat tire                   0                              113          113   
weather conditions          0                              299          299   
delayed by school           0                              267          269   
accident                    0                               32           32   

                                              created_on  borough  \
reason                      how_long_delayed                        
late return from field trip 202340                     1        1   
heavy traffic               56108                      1        1   
                            27832                      1        1   
                            26630                      1        1   
mechanical problem          25550                      1        1   
...                                                  ...      ...   
problem run                 0                         66       60   
flat tire                   0                        113      109   
weather conditions          0                        299      281   
delayed by school           0                        269      203   
accident                    0                         32       30   

                                              bus_company_name  

In [40]:
 #(vi). What was the average delay time for each reason type?

# Get the records with reasons and how long on average a delay took then sort 
# ---
# YOUR CODE GOES BELOW
avg_delay_df = df_clean.groupby('reason')['how_long_delayed'].mean()
avg_delay_df.sort_values(ascending = False)


reason
late return from field trip    62.266480
accident                       34.967933
problem run                    31.148968
mechanical problem             30.922342
won`t start                    29.644387
flat tire                      29.396742
weather conditions             29.244374
other                          28.494505
heavy traffic                  26.258358
delayed by school              19.138164
Name: how_long_delayed, dtype: float64

## 5.b) Recommendations

- Below are the recommendations from the above dataset analysis.

(i). The top most reason for delays is 'Heavy Traffic' with 173,221 occurences. The company can explore other routes with less traffic. The state government should develop more roads.
The other reasons Bus breakdowns were Mechanical Problems - 14985,
Won't start - 7731 & Flat tire - 4038.

(ii). The company should have a routine checkup or maintenance regularly the old buses should not be allowed to the road. The company can venture into frequent training of the drivers on automotive diagnosis before requesting for a mechanic.

(iii). We need to have clear routine for field trips to ensure that Boroughs pickup time and drop off its taken care of to avoid delays for the students.

## 6. Challenging your Solution

During this step, we review our solution and implement approaches that could potentially provide a better outcome. In our case, we could propose the following question that wasn't answered in our solution because it couldn't have greatly contributed to our recommendation.

In [41]:
# Which boroughs experienced the most breakdowns?

breakdowns_boro = df_clean.groupby(['borough']).count()
breakdowns_boro

# Lets Sort to get bus company with highest breakdowns

breakdowns_boro = pd.DataFrame(df_clean.groupby(['borough'])['busbreakdown_id'].count())
breakdowns_boro.sort_values(by='busbreakdown_id',ascending=0).head(10)

# we then filter to get the breakdown reasons
breakdowns_boro = pd.DataFrame(df_clean.groupby(['borough','reason'])['busbreakdown_id'].count())
breakdowns_boro.sort_values(by='busbreakdown_id',ascending=0).head(10)

busbreakdown_id
borough       reason                             
manhattan     heavy traffic                 53921
bronx         heavy traffic                 46210
brooklyn      heavy traffic                 24460
queens        heavy traffic                 16611
brooklyn      other                         16391
staten island heavy traffic                 12276
brooklyn      mechanical problem             8976
bronx         other                          8051
queens        other                          6353
              mechanical problem             6273

###  Observations

- Manhattan,bronx,brooklyn  were the top 3 in breakdown or the boroughs that experienced the most breakdowns.
- Heavy Traffic is the major reason for the delays. i recommend that this bus company should be given alternative or should be treated as essential providers.


## 7. Follow up questions

During this step, you rethink and propose other ways that you can improve your solution

### a). Did we have the right data?

- Yes we had enough data for carry out our data analysis and even for building our prediction model for future use.

### b). Do we need other data to answer our question?

- Yes we need more data for predictive models, like we would build a model regarding the number of mechanical failures, with reasons for engine model, shaft problems or even aircleaner issues.
- This will help us predict if the bus would fail or not.

### c). Did we have the right question?

- Yes we had the right questions for our analytic but more can be asked and answered.

##### Were there any other questions that we needed to have answered?

i.    Does delay in bus affect the academic status of the students?

ii.   What is the relationship between the past data and the current data that we have?

iii.  What is the corelation betwwen the weather condition and flat tire?